R/n_routine_bloodwork.r

```diff
@@ -55,58 +69,64 @@
 #'
 
 n_routine_bloodwork <- function(dbcon,
-                                cohort) {
-  cohort <- coerce_to_datatable(cohort)
-
+                                cohort,
+                                exclude_ed = FALSE) {
+  # mapping warning
   mapping_message("Sodium and Hemoglobin tests")
+  # warning to remind user of availability check
+  availability_message("lab")
 
   cat("\nThis function may take a few minutes to run...\n\n")
 
-  startwith.any <- function(x, prefix) {
-    mat <- matrix(0, nrow = length(x), ncol = length(prefix))
-    for (i in 1:length(prefix)) {
-      mat[, i] <- startsWith(x, prefix[i])
-    }
-    return(as.vector(apply(mat, MARGIN = 1, FUN = sum)) > 0)
-  }
-
-
-  temp_d_glist <- cohort$genc_id
 
-  ### query lab table:
-  DBI::dbSendQuery(dbcon, "Drop table if exists temp_data;")
-  DBI::dbWriteTable(dbcon, c("pg_temp", "temp_data"),
-    cohort[, .(genc_id)],
-    row.names = FALSE,
-    overwrite = TRUE
-  )
+  # check input type and column name
+  check_input(cohort, argtype = c("data.table", "data.frame"), colnames =  c("genc_id"))
+  check_input(exclude_ed, argtype = "logical")
+  cohort <- coerce_to_datatable(cohort)
 
   # find table name for lab table
   lab_table <- find_db_tablename(dbcon, "lab", verbose = FALSE)
 
-  lab <- DBI::dbGetQuery(
-    conn = dbcon,
+  # load lab from db
+  lab <- dbGetQuery(
+    dbcon,
     paste0(
-      "select genc_id, result_value, test_type_mapped_omop from ",
-      lab_table,
-      " where test_type_mapped_omop in ('3019550', '3000963')
-      and genc_id in (select genc_id from temp_data);"
+      ifelse(
+        exclude_ed == TRUE,
+        # filter by admission date time and exclude tests before admission
+        paste0(
+          "select l.genc_id, l.collection_date_time, l.result_value,
+           l.test_type_mapped_omop, a.admission_date_time
+           from ", lab_table, " l
+           left join admdad a
+           on l.genc_id = a.genc_id
+           where l.test_type_mapped_omop in ('3000963','3019550') and
+           l.collection_date_time >= a.admission_date_time and a.genc_id in ("
+        ),
+        # no filter on collection date time
+        paste0(
+          "select genc_id, collection_date_time, result_value, test_type_mapped_omop
+           from ", lab_table,
+           " where test_type_mapped_omop in ('3019550', '3000963') and genc_id in ("
+        )
+      ),
+      # IN method is used instead of temp table method to pull based on genc_id list,
+      # to ensure function works in HPC environment
+      paste(cohort$genc_id, collapse = ", "), ")"
     )
-  ) %>%
-    as.data.table()
-
+  ) %>% as.data.table
 
+  # only count tests with a valid result_value (numeric or starting with "<", ">")
   lab <-
     lab[, result_value := .(trimws(result_value))] %>%
-    .[!is.na(as.numeric(result_value)) | startwith.any(result_value, c("<", ">"))] %>%
+    .[!is.na(as.numeric(result_value)) | grepl("^([<>])", result_value)] %>%
```

Comment:
Good point. And we have the same inconsistency issue in MPR. Currently in MPR the bloodwork indicator does not filter by whether the result is numeric or not, it counts all results. For the RBC transfusion indicator, it takes numeric values, but I just found that the code does not do what it claims to do:
```
# keep numeric values or those starting with <, >, @
  # exclude other non-numeric hgb result values
  hemoglobin[, result_value := as.numeric(
    stringr::str_split(result_value,pattern = "@|<|>",
                       simplify = TRUE)[,1])]
``` 
This code will keep any value before "<" or ">" or "@", which is right only for "@". For "<" or ">", we want to include the value following "<" or ">", not after. 

Here is what I do for my research projects and I just had a look at all Sodium and HGB test results it would capture most of the non-num result values except for some very rare edge cases. 
Generally the following 2 types should be included:
1. starting with ">", "<" or "LESS THAN", and followed by a numeric number
2. starting with a numeric number, followed by some string starting with "@", like `"138@WPDR"`, `"84@WPDR"`, `"129@"`. In this type we should only include the number before the "@" sign, not after. Results like `"TO CT @2137"`, "TO ER @1720" should be excluded.
In some rare cases it is the combination of the two, like `">180@"`.

I will fix the two functions here following this approach and create a new issue in MPR repo.

Reply:
Amazing, thanks so much @guoyi-smh for digging into this! That all makes sense to me, and thanks so much for ensuring consistency between MPR/OPR and Rgemini. I reviewed the updated code and everything looks good. I'll go ahead and resolve this conversation.